rts/motoko-rts-tests/src/gc.rs

```diff
@@ -0,0 +1,211 @@
+// Naming conventions:
+//
+// - offset = index in the "heap" array/slice/vector
+// - address = address in the process's address space
+//
+// To convert an offset into an address, add heap array's address to the offset.
+
+mod heap;
+mod utils;
+
+use heap::MotokoHeap;
+use utils::{read_word, ObjectIdx, GC, GC_IMPLS, WORD_SIZE};
+
+// use motoko_rts::debug;
+use motoko_rts::gc::copying::copying_gc_internal;
+use motoko_rts::gc::mark_compact::compacting_gc_internal;
+use motoko_rts::types::*;
+
+use std::collections::{BTreeMap, HashMap, HashSet};
+
+pub fn test() {
+    println!("Testing garbage collection ...");
+
+    // TODO: Add more tests
+
+    let heap = btreemap! {
+        0 => vec![0, 2],
+        2 => vec![0],
+        3 => vec![3],
+    };
+
+    let roots = vec![0, 2, 3];
+
+    test_gcs(&TestHeap { heap, roots });
+}
+
+#[derive(Debug)]
+struct TestHeap {
+    heap: BTreeMap<ObjectIdx, Vec<ObjectIdx>>,
+    roots: Vec<ObjectIdx>,
+}
+
+/// Test all GC implementations with the given heap
+fn test_gcs(heap_descr: &TestHeap) {
+    for gc in &GC_IMPLS {
+        test_gc(*gc, &heap_descr.heap, &heap_descr.roots);
+    }
+}
+
+fn test_gc(gc: GC, refs: &BTreeMap<u32, Vec<u32>>, roots: &[u32]) {
+    let heap = MotokoHeap::new(refs, roots, gc);
+
+    // println!("{:?}", heap.heap);
+
+    // unsafe {
+    //     debug::dump_heap(
+    //         heap.heap_base_address() as u32,
+    //         heap.heap_ptr_address() as u32,
+    //         skew(heap.static_root_array_address()),
+    //         heap.closure_table_address() as *mut SkewedPtr,
+    //     );
+    // }
+
+    // Check `check_dynamic_heap` sanity
+    check_dynamic_heap(
+        refs,
+        &roots,
+        &**heap.heap(),
+        heap.heap_base_offset(),
+        heap.heap_ptr_offset(),
+    );
+
+    for _ in 0..3 {
+        gc.run(heap.clone());
+
+        let heap_base_offset = heap.heap_base_offset();
+        let heap_ptr_offset = heap.heap_ptr_offset();
+        check_dynamic_heap(
+            refs,
+            &roots,
+            &**heap.heap(),
+            heap_base_offset,
+            heap_ptr_offset,
+        );
+    }
+}
+
+/// Check the dynamic heap:
+///
+/// - Objects should point to objects with the expected tags (as specified by the `objects` argument)
+/// - Each tag should be seen at most once
+/// - All of `roots` should be seen
```

Comment:
@ulan points out that we should check that we only see the transitive closure of roots here.